### PR TITLE
[chainerx] add `logical_and` `logical_or`

### PR DESCRIPTION
**Pull request by kshitij12345**
Adds `logical_and` and `logical_or`

References : #6423

### CODE REVIEW

**Comment by kshitij12345** [comment]
Failure looks irrelevant. Please Review.

File: chainerx_cc/chainerx/cuda/cuda_device/comparison.cu [hunk 1]
```diff
@@ -135,6 +135,54 @@ public:
 
 CHAINERX_REGISTER_OP_CUDA(LogicalNotOp, CudaLogicalNotOp);
 
+template <typename T>
+struct LogicAndImpl {
```
Comment (hvy) [inline chainerx_cc/chainerx/cuda/cuda_device/comparison.cu]: A minor but could you align naming convention to `LogicalNotImpl`, i.e.
```suggestion
struct LogicalAndImpl {
```

File: tests/chainerx_tests/unit_tests/routines_tests/test_logic.py [hunk 2]
```diff
@@ -205,3 +205,105 @@ def forward_xp(self, inputs, xp):
...
+    else:
+        return a.astype(bool, True)
+
+
+def logical_and_lambda(xp, a, b):
+    return xp.logical_and(a, b)
+
+
+def logical_or_lambda(xp, a, b):
+    return xp.logical_and(a, b)
```
Comment (hvy) [inline tests/chainerx_tests/unit_tests/routines_tests/test_logic.py]: Ops, this should be the following?
```suggestion
    return xp.logical_or(a, b)
```

File: tests/chainerx_tests/unit_tests/routines_tests/test_logic.py [hunk 3]
```diff
@@ -205,3 +205,105 @@ def forward_xp(self, inputs, xp):
...
+        a_dtype, b_dtype = self.dtypes
+        a = numpy.array(a_object, a_dtype)
+        b = numpy.array(b_object, b_dtype)
+        return a, b
+
+    def forward_xp(self, inputs, xp):
+        a, b = inputs
+
+        def on_cuda(a):
+            return 'cuda' in str(a.device)
```
Comment (hvy) [inline tests/chainerx_tests/unit_tests/routines_tests/test_logic.py]: You can get the backend name "cuda" via `a.device.backend.name`.

**Comment by hvy** [comment]
Thanks for the PR as always. I wrote some comments if you could take a look. Maybe we should wait until https://github.com/chainer/chainer/pull/6780 is merged.

File: chainerx_cc/chainerx/cuda/cuda_device/comparison.cu (hunk 1 above)
Comment (kshitij12345) [reply chainerx_cc/chainerx/cuda/cuda_device/comparison.cu]: Oh yes. Missed that.

File: tests/chainerx_tests/unit_tests/routines_tests/test_logic.py (hunk 2 above)
Comment (kshitij12345) [reply tests/chainerx_tests/unit_tests/routines_tests/test_logic.py]: Oh right.

File: tests/chainerx_tests/unit_tests/routines_tests/test_logic.py (hunk 3 above)
Comment (kshitij12345) [reply tests/chainerx_tests/unit_tests/routines_tests/test_logic.py]: Since #6780  was merged, will remove this

**Comment by kshitij12345** [comment]
Have updated the code as per the review and also removed the hack.

File: tests/chainerx_tests/unit_tests/routines_tests/test_logic.py (hunk 3 above)
Comment (kshitij12345) [reply tests/chainerx_tests/unit_tests/routines_tests/test_logic.py]: Btw, thanks for the suggestion. Will keep that in mind for future.

File: tests/chainerx_tests/unit_tests/routines_tests/test_logic.py (hunk 2 above)
Comment (hvy) [reply tests/chainerx_tests/unit_tests/routines_tests/test_logic.py]: This still seems to be an issue. Could you recheck it?

**Comment by hvy** [comment]
One of the comments seems unaddressed. Could you fix it? https://github.com/chainer/chainer/pull/6779#discussion_r273324075

**Comment by kshitij12345** [comment]
Haha, totally missed that silly typo. Thanks. Have updated.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Test failures seem unrelated.

**Comment by hvy** [comment]
LGTM!

**Pull request merged**